test/e2e/kcp_remediations.go

```diff
@@ -0,0 +1,693 @@
+	lister                          framework.Lister
+	namespace                       string
+	clusterName                     string
+	expectedReplicas                int
+	expectedOldMachines             []string
+	expectedDeletedMachines         []string
+	waitForMachinesIntervals        []interface{}
+	checkMachineListStableIntervals []interface{}
+}
+
+// waitForMachines waits for machines to reach a well known state defined by number of replicas, a list of machines to exists,
+// a list of machines to not exists anymore. The func also check that the state is stable for some time before
+// returning the list of new machines.
+func waitForMachines(ctx context.Context, input waitForMachinesInput) (allMachineNames, newMachineNames []string) {
+	inClustersNamespaceListOption := client.InNamespace(input.namespace)
+	matchClusterListOption := client.MatchingLabels{
+		clusterv1.ClusterNameLabel:         input.clusterName,
+		clusterv1.MachineControlPlaneLabel: "",
+	}
+
+	expectedOldMachines := sets.NewString(input.expectedOldMachines...)
+	expectedDeletedMachines := sets.NewString(input.expectedDeletedMachines...)
+	allMachines := sets.NewString()
+	newMachines := sets.NewString()
+	machineList := &clusterv1.MachineList{}
+
+	// Waits for the desired set of machines to exist.
+	log.Logf("Waiting for %d machines, must have %s, must not have %s", input.expectedReplicas, expectedOldMachines.List(), expectedDeletedMachines.List())
+	Eventually(func() bool {
+		// Gets the list of machines
+		if err := input.lister.List(ctx, machineList, inClustersNamespaceListOption, matchClusterListOption); err != nil {
+			return false
+		}
+		allMachines = sets.NewString()
+		for i := range machineList.Items {
+			allMachines.Insert(machineList.Items[i].Name)
+		}
+
+		// Compute new machines (all - old - to be deleted)
+		newMachines = allMachines.Clone()
+		newMachines.Delete(expectedOldMachines.List()...)
+		newMachines.Delete(expectedDeletedMachines.List()...)
+
+		log.Logf(" - expected %d, got %d: %s, of which new %s, must have check: %t, must not have check: %t", input.expectedReplicas, allMachines.Len(), allMachines.List(), newMachines.List(), allMachines.HasAll(expectedOldMachines.List()...), !allMachines.HasAny(expectedDeletedMachines.List()...))
+
+		// Ensures all the expected old machines are still there.
+		if !allMachines.HasAll(expectedOldMachines.List()...) {
+			return false
+		}
+
+		// Ensures none of the machines to be deleted is still there.
+		if allMachines.HasAny(expectedDeletedMachines.List()...) {
+			return false
+		}
+
+		return allMachines.Len() == input.expectedReplicas
+	}, input.waitForMachinesIntervals...).Should(BeTrue(), "Failed to get the expected list of machines: got %s (expected %d machines, must have %s, must not have %s)", allMachines.List(), input.expectedReplicas, expectedOldMachines.List(), expectedDeletedMachines.List())
+	log.Logf("Got %d machines: %s", input.expectedReplicas, allMachines.List())
```

Comment:
This seems wrong `Got (expectedReplicas)`

Reply:
If we reach this point we actually got expectedReplicas, but changed to 
```suggestion
	log.Logf("Got %d machines: %s", allMachines.Len(), allMachines.List())
```